src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -165,29 +179,194 @@ class DecisionTree(BaseModel):
         if not self.predictor:
             raise ValueError("The tree has not been trained before trying to predict")
         return self.predictor.predict_leaf_matrix(X, scale)
 
+    def __remove_leaf_nodes(self) -> None:
+        cdef:
+            int i, n_nodes
+            object parent
+        n_nodes = len(self.leaf_nodes)
+        for i in range(n_nodes):
+            parent = self.leaf_nodes[i].parent
+            if parent.left_child == self.leaf_nodes[i]:
+                parent.left_child = None
+            else:
+                parent.right_child = None
+            self.leaf_nodes[i] = None
+
+    def __fit_new_leaf_nodes(self, X: np.ndarray, Y: np.ndarray, sample_weight:
+                             np.ndarray, indices: np.ndarray) -> None:
+        cdef:
+            int idx, n_obs, depth, cur_split_idx
+            double cur_threshold
+            object cur_node
+            int[::1] all_idx
+            int[::1] leaf_indices
+
+        refit_objs = []
+        n_obs = indices.shape[0]
+        for idx in indices:
+            cur_node = self.root
+            depth = 0
+            while isinstance(cur_node, DecisionNode) :
+                # Mark cur_node as visited
+                cur_node.visited = 1
+                cur_split_idx = cur_node.split_idx
+                cur_threshold = cur_node.threshold
+
+                # Check if X should go to the left or right
+                if X[idx, cur_split_idx] < cur_threshold:
+                    # If the left or right is none, then there previously was a
+                    # leaf node, and we create a new refit object
+                    if cur_node.left_child is None:
+                        cur_node.left_child = refit_object(idx, depth,
+                                                           cur_node, True)
+                        refit_objs.append(cur_node.left_child)
+                    # If there already is a refit object, add this new index to
+                    # the refit object
+                    elif isinstance(cur_node.left_child, refit_object):
+                        cur_node.left_child.add_idx(idx)
+
+                    cur_node = cur_node.left_child
+                else:
+                    if cur_node.right_child is None:
+                        cur_node.right_child = refit_object(idx, depth,
+                                                            cur_node, False)
+                        refit_objs.append(cur_node.right_child)
+                    elif isinstance(cur_node.right_child, refit_object):
+                        cur_node.right_child.add_idx(idx)
+
+                    cur_node = cur_node.right_child
+                depth += 1
+
+        all_idx = np.arange(0, indices.shape[0], dtype=np.int32)
+        leaf_builder = self.leaf_builder_class(X, Y, all_idx)
+        criteria = self.criteria_class(X, Y, sample_weight)
+        # Make refit objects into leaf_nodes
+        n_obs = len(refit_objs)
```

Comment:
n_obs is the wrong name for this or not?

Reply:
As this is a local variable not used outside the scope of this function, the general idea was to just reuse the same variable tat is initialized previously as to not use another one. However, it is correct that n_objs would be more describing.